### PR TITLE
Add Request ID Tracking for Enhanced Debugging and Distributed Tracing

### PR DESCRIPTION
**Pull request by Drjacky**
## Problem
Currently, the SDK lacks request correlation capabilities, making it difficult to:
- Debug production issues by correlating errors with specific API requests
- Track requests across distributed systems and microservices
- Provide specific request identifiers to MongoDB support for faster resolution

## Solution
Implement request ID propagation that enables distributed tracing through:
- Automatic `X-Request-ID` header inclusion from context
- Enhanced error messages with request context
- Zero-breaking changes with opt-in functionality

### Template and Specification Changes (Source)
- **`openapi/atlas-api.yaml`** - Added `x-is-api-error: true` vendor extension to ApiError schema
- **`tools/config/go-templates/client.mustache`** - Added request ID header propagation in `prepareRequest` method
- **`tools/config/go-templates/model_simple.mustache`** - Added `GetContextualError` method with context imports

### Generated Code (Expected Output)
- **`admin/client.go`** - Updated `prepareRequest` method with request ID handling
- **`admin/model_api_error.go`** - Added `GetContextualError` method to ApiError struct

### Tests
- **`admin/client_test.go`** - Tests for request ID header propagation
- **`admin/model_api_error_test.go`** - Tests for `GetContextualError` method functionality

### Basic Usage
```go
// Set request ID in context
ctx := context.WithValue(context.Background(), "X-Request-ID", "support-ticket-456")

// All API calls automatically include the request ID header
project, _, err := client.ProjectsApi.GetProject(ctx, "project-id")

// Enhanced error messages include request ID for correlation
if err != nil {
    if apiErr, ok := err.(*admin.ApiError); ok {
        log.Printf("Error: %s", apiErr.GetContextualError(ctx))
        // Output: "404 RESOURCE_NOT_FOUND Project not found (RequestID: support-ticket-456)"
    }
}
```

### Distributed Tracing Example
```go
// In microservice architecture
func handleRequest(w http.ResponseWriter, r *http.Request) {
    // Propagate request ID through your call stack
    requestID := r.Header.Get("X-Request-ID")
    if requestID == "" {
        requestID = generateRequestID()
    }
    
    ctx := context.WithValue(r.Context(), "X-Request-ID", requestID)
    
    // All MongoDB API calls will include the same request ID
    cluster, _, err := client.ClustersApi.CreateCluster(ctx, clusterReq)
    
    if err != nil {
        // Return request ID to client for support tickets
        w.Header().Set("X-Request-ID", requestID)
        w.WriteHeader(500)
        json.NewEncoder(w).Encode(map[string]string{
            "error": "Operation failed",
            "request_id": requestID, // Client can reference this with support
        })
    }
}
```

### Benefits
- Production Debugging: Correlate errors with specific requests in logs
- Distributed Tracing: Track requests across microservices and systems
- Support Efficiency: Provide specific request IDs to MongoDB support
- Monitoring: Enhanced observability for enterprise deployments
- Zero Breaking Changes: Fully optional and backward compatible

### Generation Note
This PR modifies source templates and specifications. The included Go file changes reflect the expected output from regeneration. After merging, the following regeneration steps are recommended:

1. Regenerate openapi/atlas-api-transformed.yaml from atlas-api.yaml
2. Regenerate Go client files from templates and transformed specification

The changes are designed to work with the existing generation pipeline in tools/scripts/generate.sh.

### Testing
Comprehensive tests included:

- Request ID header propagation in HTTP client
- Error message enhancement with request context
- Edge cases (empty request IDs, non-string values)
- Backward compatibility verification

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: admin/model_api_error.go [hunk 1]
```diff
@@ -219,3 +219,14 @@ func (o *ApiError) HasReason() bool {
...
+func (a *ApiError) GetContextualError(ctx context.Context) string {
+    baseError := a.Error()
+
+    // Check if context has request ID
+    if requestID, ok := ctx.Value("X-Request-ID").(string); ok && requestID != "" {
+        return fmt.Sprintf("%s (RequestID: %s)", baseError, requestID)
+    }
+
+    return baseError
+}
```
Comment (AgustinBettati) [inline admin/model_api_error.go]: Is there a reason we want to place this utility logic within ApiError model? Potentially this can be simplified to a utility function separate from ApiError which simply obtains the RequestID from the Go context.

**Review by AgustinBettati — COMMENTED** [review]
Hi @Drjacky, thank you for the details and proposed changes. 

I would like to gain more context on what value we gain from adding a client defined request-id header to Atlas API. We can continue conversation directly in each comment.

**Comment by Drjacky** [comment]
> Hi @Drjacky, thank you for the details and proposed changes.
> 
> I would like to gain more context on what value we gain from adding a client defined request-id header to Atlas API. We can continue conversation directly in each comment.

The main value is enabling better distributed tracing between client applications and the Atlas API. This allows developers to:

- Correlate client-side errors with server-side logs
- Trace requests across service boundaries for debugging
- Provide specific request IDs to support for faster issue resolution
- Follow industry-standard practices for request correlation in distributed systems

This is particularly valuable for enterprise applications and complex deployments where understanding the full request flow is crucial for debugging and monitoring.

**Comment by AgustinBettati** [comment]
With latest inputs I am seeing the following 2 aspects being handled in this PR:

1. Include "X-Request-ID" header in Atlas API requests if value is present in Go context.
2. Provide utility function `GetRequestID` for obtaining the request id value from a Go context [related comment](https://github.com/mongodb/atlas-sdk-go/pull/674#discussion_r2479040921).

For `1` I would not merge this change until we have confirmation from support team that these headers can provide actual value for debugging. Would advise contacting Atlas support directly through their portal https://support.mongodb.com/.

For `2` `GetRequestID` utility function seems like a more general case function that is not specific to the Atlas Go SDK.

With these considerations I would opt for closing this PR as of now until we receive further inputs from support. Happy to continue the conversation.
